atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultBucketAssigner.java

```diff
@@ -0,0 +1,235 @@
+                            + " we're very far behind, or because there's a bug in the bucket creation logic that"
+                            + " causes it to infinitely create the same bucket or return CLOSED over and over again."
+                            + " The last bucket processed has ID {} and reached state {}",
+                        SafeArg.of("bucketIdentifier", iterationState.bucketIdentifier()),
+                        SafeArg.of("bucketAssignerState", iterationState.bucketAssignerState()));
+            }
+        }
+    }
+
+    private IterationState createNewBucket() {
+        BucketStateAndIdentifier stateAndIdentifier =
+                sweepBucketAssignerStateMachineTable.getBucketStateAndIdentifier();
+        BucketStateVisitor visitor = new BucketStateVisitor(stateAndIdentifier.bucketIdentifier());
+        IterationState state = stateAndIdentifier.state().accept(visitor);
+        metrics.bucketStateForIteration(state.iterationState());
+        log.info(
+                "Processed a bucket with ID {} and reached state {}",
+                SafeArg.of("bucketIdentifier", state.bucketIdentifier()),
+                SafeArg.of("bucketAssignerState", state.bucketAssignerState()));
+        return state;
+    }
+
+    private final class BucketStateVisitor implements BucketAssignerState.Visitor<IterationState> {
+        private final long bucketIdentifier;
+
+        private BucketStateVisitor(long bucketIdentifier) {
+            this.bucketIdentifier = bucketIdentifier;
+        }
+
+        @Override
+        public IterationState visit(BucketAssignerState.Start start) {
+            metrics.progressedToBucketIdentifier(bucketIdentifier);
+            metrics.bucketAssignerState(start);
+            OptionalLong closeTimestamp = closeTimestampCalculator.apply(start.startTimestampInclusive());
+
+            if (closeTimestamp.isPresent()) {
+                return tryTransitionAndExecuteNextState(
+                        start,
+                        BucketAssignerState.immediatelyClosing(
+                                start.startTimestampInclusive(), closeTimestamp.orElseThrow()));
+            } else {
+                return tryTransitionAndExecuteNextState(
+                        start, BucketAssignerState.opening(start.startTimestampInclusive()));
+            }
+        }
+
+        @Override
+        public IterationState visit(BucketAssignerState.Opening open) {
+            metrics.bucketAssignerState(open);
+            WriteState writeState = bucketWriter.writeToAllBuckets(
+                    bucketIdentifier, Optional.empty(), TimestampRange.openBucket(open.startTimestampInclusive()));
+
+            if (writeState == WriteState.FAILED_CAS) {
+                return IterationState.of(State.CONTENTION_ON_WRITES, bucketIdentifier, open);
+            }
+
+            BucketAssignerState waitingUntilCloseable =
```

Comment:
thought: If we get here, the write must have been SUCCESS (a failure to do the write themselves that wasn't a FAILED_CAS would have thrown an exception).